### PR TITLE
drop root privileges and add network capabilities for blackbox_exporter process

### PR DESCRIPTION
**Pull request by othmane399**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -8,12 +8,22 @@ RUN apk add make git && \
...
+RUN apk add -U libcap
 
-COPY --from=0 /go/src/github.com/prometheus/blackbox_exporter/blackbox_exporter  /bin/blackbox_exporter
+
+RUN cp blackbox_exporter  /bin/blackbox_exporter
 COPY blackbox.yml       /etc/blackbox_exporter/config.yml
 
+ARG USERNAME=blackbox_exporter
+
+RUN setcap cap_net_raw+ep /bin/blackbox_exporter
```
Comment (bpaquet) [inline Dockerfile]: What is the impact of that on kubernetes ?

Comment (othmane399) [reply Dockerfile]: All the process will run under a non-root uid but with network capabilities for blackbox_exporter binary to work properly

**Comment by othmane399** [comment]
Ok for debian !

**Comment by othmane399** [comment]
> Can you work with the devops team and move it to debian ?

Ok for debian !

**Comment by bpaquet** [comment]
Can we set that in kube config instead of here ? If yes, what is the policy
?

File: Dockerfile [hunk 2]
```diff
@@ -1,25 +1,29 @@
-FROM golang:alpine
+FROM debian:stretch-slim
```
Comment (bpaquet) [inline Dockerfile]: Please use the base image.

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,33 @@
...
-WORKDIR /go/src/github.com/prometheus/blackbox_exporter
+WORKDIR /go/src/github.com/prometheus/
 
-COPY . .
+RUN apt-get update && apt-get dist-upgrade -y && apt-get clean
 
-RUN apk add make git && \
-    go get && \
-    make build
+RUN apt-get update && apt-get install make git libcap2-bin golang-1.8-go  -y && apt-get clean
```
Comment (bpaquet) [inline Dockerfile]: Please use a multistage build to not deploy all the build stuff in production

File: Dockerfile (hunk 2 above)
Comment (othmane399) [reply Dockerfile]: debian:stretch-slim is not the base image ? 
https://github.com/doctolib/docker-images/blob/master/static/base/Dockerfile

**Comment by bpaquet** [comment]
https://github.com/doctolib/docker-images/blob/master/non_static/postgres/Dockerfile

File: Dockerfile [hunk 4]
```diff
@@ -1,19 +1,43 @@
...
+FROM ${ROOT_IMAGE}:${BASE_VERSION}
+
+
+WORKDIR /go/src/github.com/prometheus/
 
-WORKDIR /go/src/github.com/prometheus/blackbox_exporter
 
-COPY . .
+RUN apt-get update && apt-get install make git golang-1.8-go  -y && apt-get clean
+
```
Comment (bpaquet) [inline Dockerfile]: Empty line

File: Dockerfile [hunk 5]
```diff
@@ -1,19 +1,43 @@
...
+
+RUN git clone https://github.com/doctolib/blackbox_exporter.git --depth=1
```
Comment (bpaquet) [inline Dockerfile]: ??? You should use COPY . .

File: Dockerfile [hunk 6]
```diff
@@ -1,19 +1,37 @@
-FROM golang:alpine
+ARG BASE_VERSION=20190221_114834
+ARG ROOT_IMAGE=580698825394.dkr.ecr.eu-central-1.amazonaws.com/base
+
+FROM ${ROOT_IMAGE}:${BASE_VERSION}
 
 WORKDIR /go/src/github.com/prometheus/blackbox_exporter
 
+RUN apt-get update && apt-get install make git golang-1.8-go  -y && apt-get clean
```
Comment (bpaquet) [inline Dockerfile]: Remove git

Comment (othmane399) [reply Dockerfile]: `go get` need `git` to compile :
```
go: missing Git command. See https://golang.org/s/gogetcmd
package github.com/prometheus/promu: exec: "git": executable file not found in $PATH
```
but it's not in the 2nd stage so i think it's fine !

**Pull request merged**